packages/big-design/src/components/Tree/TreeNode/TreeNode.tsx

```diff
@@ -0,0 +1,328 @@
+      }
+    }
+  };
+
+  const handleNodeSelected = useCallback(() => {
+    if (selectable === undefined || value === undefined || disabled) {
+      return;
+    }
+
+    let newSelectedValues: T[] = [];
+
+    if (selectable === 'multi') {
+      if (state.selectedValues.includes(value)) {
+        newSelectedValues = state.selectedValues.filter((selectedValue) => selectedValue !== value);
+      } else {
+        newSelectedValues = [...state.selectedValues, value];
+      }
+
+      if (typeof onSelect === 'function') {
+        onSelect(newSelectedValues);
+      }
+    }
+
+    if (selectable === 'radio') {
+      newSelectedValues = [value];
+
+      if (typeof onSelect === 'function') {
+        onSelect(newSelectedValues[0]);
+      }
+    }
+
+    dispatch({ type: 'SELECTED_NODE', values: newSelectedValues });
+  }, [disabled, dispatch, onSelect, selectable, state.selectedValues, value]);
+
+  // Needs to handle the following keyboard events:
+  // https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-22
+  const handleKeyEvent = (e: React.KeyboardEvent<HTMLLIElement>) => {
+    const key = e.key;
+
+    if (e.altKey || e.currentTarget !== e.target) {
+      return;
+    }
+
+    switch (key) {
+      case ' ':
+        e.preventDefault();
+
+        handleNodeSelected();
+        break;
+
+      case 'Enter':
+        if (nodeRef.current === e.currentTarget) {
+          if (thisRef.current.children) {
+            handleNodeToggle();
+          } else if (selectable) {
+            handleNodeSelected();
+          }
+        }
+        break;
+
+      case 'ArrowDown':
+        e.preventDefault();
+
+        dispatch({ type: 'FOCUS_DOWN', id: state.focusedNode });
+        break;
+
+      case 'ArrowUp':
+        e.preventDefault();
+
+        dispatch({ type: 'FOCUS_UP', id: state.focusedNode });
+        break;
+
+      case 'ArrowRight':
+        dispatch({ type: 'FOCUS_RIGHT', id: state.focusedNode });
+        break;
+
+      case 'ArrowLeft':
+        dispatch({ type: 'FOCUS_LEFT', id: state.focusedNode });
+        break;
+
+      case 'Home':
+        dispatch({ type: 'FOCUS_FIRST' });
+        break;
+
+      case 'End':
+        dispatch({ type: 'FOCUS_LAST' });
+        break;
+
+      default:
+        break;
+    }
+  };
+
+  const additionalProps = useMemo(() => (selectable ? { 'aria-selected': selected } : {}), [selectable, selected]);
+
+  const renderedArrow = useMemo(
+    () =>
+      thisRef.current.children ? (
+        <StyledArrowWrapper expanded={expanded} flexShrink={0}>
+          <ChevronRightIcon color="secondary60" focusable={false} size="xLarge" />
+        </StyledArrowWrapper>
+      ) : (
+        <StyledGap />
+      ),
+    [expanded],
+  );
+
+  const renderedChildren = useMemo(
+    () =>
+      thisRef.current.children &&
+      expanded && (
+        <StyledUl role="group">
+          {isLoading && thisRef.current.children.length < 1 ? (
+            <TreeNode
+              id={-1}
```

Comment:
The `-1` id is only for the loading state node. As you can see I don't pass any of the callback handlers into the `TreeNode` so it should be safe. The only thing I would worry about in this situation is having that id being passed into the underlying state, but luckily it will never do that. This id will also never be passed as an HTML id attribute so we shouldn't need to be concerned about uniqueness.

Reply:
Ah I see, so it's just the loading state node. If that's the case, I think it's fine. Also I just realise `id` is not passed to the `onSelect` callback. I wonder if we should (please see my comment below) 🤔